src/VisualStudio/Core/Def/Packaging/PackageSearchService.cs

```diff
@@ -0,0 +1,752 @@
+                        // File was downloaded.  
+                        return resultOpt;
+                    }
+                }
+            }
+        }
+
+        /// <summary>Returns 'null' if download is not available and caller should keep polling.</summary>
+        private async Task<XElement> TryDownloadFileAsync(IPackageSearchRemoteControlClient client)
+        {
+            LogInfo("Read file from client");
+
+            // "ReturnsNull": Only return a file if we have it locally *and* it's not older than our polling time (1 day).
+
+            using (var stream = await client.ReadFileAsync(__VsRemoteControlBehaviorOnStale.ReturnsNull).ConfigureAwait(false))
+            {
+                if (stream == null)
+                {
+                    LogInfo("Read file completed. Client returned no data");
+                    return null;
+                }
+
+                LogInfo("Read file completed. Client returned data");
+                LogInfo("Converting data to XElement");
+                var result = XElement.Load(stream);
+                LogInfo("Converting data to XElement completed");
+                return result;
+            }
+        }
+
+        private async Task RepeatIOAsync(Action action)
+        {
+            const int repeat = 6;
+            for (var i = 0; i < repeat; i++)
+            {
+                _cancellationToken.ThrowIfCancellationRequested();
+
+                try
+                {
+                    action();
+                    return;
+                }
+                catch (Exception e) when (_swallowException(e))
+                {
+                    var delay = _delayService.FileWriteDelay;
+                    LogException(e, $"Operation failed. Trying again after {delay}");
+                    await Task.Delay(delay, _cancellationToken).ConfigureAwait(false);
+                }
+            }
+        }
+
+        private byte[] ParseDatabaseElement(XElement element)
+        {
+            LogInfo("Parsing database element");
+            var contentsAttribute = element.Attribute(ContentAttributeName);
+            if (contentsAttribute == null)
+            {
+                throw new FormatException($"Database element invalid. Missing '{ContentAttributeName}' attribute");
+            }
+
+            var text = contentsAttribute.Value;
+            var compressedBytes = Convert.FromBase64String(text);
+
+            using (var inStream = new MemoryStream(compressedBytes))
+            using (var deflateStream = new DeflateStream(inStream, CompressionMode.Decompress))
+            using (var outStream = new MemoryStream())
+            {
+                deflateStream.CopyTo(outStream);
+                var bytes = outStream.ToArray();
+
+                LogInfo($"Parsing complete. bytes.length={bytes.Length}");
+                return bytes;
+            }
+        }
+
+        public IEnumerable<PackageWithTypeResult> FindPackagesWithType(
```

Comment:
Would it make sense to separate out this things public API parts from the messy implementation of caching things on the filesystem (at least a different file maybe?)